# PostgreSQL


In [60]:
import psycopg2 as ps

In [18]:
'''
credenciales = {
    'dbname':'peliculas',
    'user':'postgres',
    'password':'tkmcardona',
    'host':'localhost',
    'port':5432
}
conexion = ps.connect( **credenciales )
'''


"\ncredenciales = {\n    'dbname':'peliculas',\n    'user':'postgres',\n    'password':'tkmcardona',\n    'host':'localhost',\n    'port':5432\n}\nconexion = ps.connect( **credenciales )\n"

In [61]:
conexion = ps.connect('dbname=peliculas user = postgres password=tkmcardona host=localhost port=5432')

In [62]:
# Crea un objeto cursor utilizando el objeto de conexión devuelto por el método connect 
# para ejecutar consultas PostgreSQL desde Python.

cursor = conexion.cursor()
cursor

<cursor object at 0x000001DFB123DC80; closed: 0>

In [63]:
sql = """
    SELECT pa.pais, ci.ciudad
    FROM ciudades ci INNER JOIN paises pa ON (ci.pais_id = pa.pais_id)
    ORDER BY pa.pais ASC;
"""
# Los métodos execute() ejecutan la consulta SQL y devuelven el resultado.

cursor.execute( sql )

for registro in cursor:
    print( registro )

('Afghanistan', 'Kabul')
('Algeria', 'Bchar')
('Algeria', 'Batna')
('Algeria', 'Skikda')
('American Samoa', 'Tafuna')
('Angola', 'Namibe')
('Angola', 'Benguela')
('Anguilla', 'South Hill')
('Argentina', 'Escobar')
('Argentina', 'Ezeiza')
('Argentina', 'Quilmes')
('Argentina', 'Crdoba')
('Argentina', 'Almirante Brown')
('Argentina', 'Merlo')
('Argentina', 'Avellaneda')
('Argentina', 'Baha Blanca')
('Argentina', 'La Plata')
('Argentina', 'Santa F')
('Argentina', 'Tandil')
('Argentina', 'Vicente Lpez')
('Argentina', 'San Miguel de Tucumn')
('Armenia', 'Yerevan')
('Australia', 'Woodridge')
('Austria', 'Salzburg')
('Austria', 'Graz')
('Austria', 'Linz')
('Azerbaijan', 'Sumqayit')
('Azerbaijan', 'Baku')
('Bahrain', 'al-Manama')
('Bangladesh', 'Dhaka')
('Bangladesh', 'Jamalpur')
('Bangladesh', 'Tangail')
('Belarus', 'Mogiljov')
('Belarus', 'Molodetno')
('Bolivia', 'Sucre')
('Bolivia', 'El Alto')
('Brazil', 'Luzinia')
('Brazil', 'Boa Vista')
('Brazil', 'Santa Brbara dOeste')
('Brazil', 'Santo 

## pandas

In [65]:
import pandas as pd

df = pd.read_sql(sql, conexion)
df

pais      ciudad
0       Afghanistan       Kabul
1           Algeria       Bchar
2           Algeria       Batna
3           Algeria      Skikda
4    American Samoa      Tafuna
..              ...         ...
595           Yemen        Aden
596           Yemen       Taizz
597      Yugoslavia  Kragujevac
598      Yugoslavia    Novi Sad
599          Zambia       Kitwe

[600 rows x 2 columns]

In [67]:
sql = """
    SELECT cl.cliente_id, cl.nombre, cl.apellido, SUM(pa.cantidad) as importe
    FROM pagos as pa INNER JOIN clientes as cl ON (pa.cliente_id = cl.cliente_id)
    GROUP BY cl.cliente_id, cl.nombre, cl.apellido
    HAVING SUM(pa.cantidad) > 100
    ORDER BY importe DESC;
"""

df = pd.read_sql(sql, conexion)
df

cliente_id   nombre  apellido  importe
0           148  Eleanor      Hunt   211.55
1           526     Karl      Seal   208.58
2           178   Marion    Snyder   194.61
3           137   Rhonda   Kennedy   191.62
4           144    Clara      Shaw   189.60
..          ...      ...       ...      ...
291         517     Brad   Mccurdy   100.76
292         572   Sidney  Burleson   100.76
293         251   Vickie    Brewer   100.75
294         219   Willie    Howell   100.75
295         135  Juanita     Mason   100.72

[296 rows x 4 columns]

In [70]:
# utilizar los métodos cursor.close() y connection.close() para cerrar las conexiones 
# PostgreSQL una vez finalizado el trabajo
cursor.close()
conexion.close()


# Facturacion

In [73]:
# CODE HERE
credenciales = {
    'dbname':'facturacion',
    'user':'postgres',
    'password':'tkmcardona',
    'host':'localhost',
    'port':5432
}

conexion = ps.connect( **credenciales )
cursor = conexion.cursor()

In [40]:
# Visualizar la tabla producto 
sql = """
    SELECT *
    FROM producto;
"""
cursor.execute( sql )

for registro in cursor:
    print( registro )

(1, 'Coca Cola (Original) - 3lt.', '$3.00')
(2, 'Coca Cola (Original) - 2lt.', '$2.00')
(3, 'Coca Cola (Original) - 1lt.', '$1.00')
(4, 'Poni Malta - 311cm3 Retornable', '$1.00')
(5, 'Poni Malta - 225cm3 Retornable', '$0.60')
(7, 'Poni Malta - 330cm3 PET Descartable', '$0.80')
(8, 'Pilsener 600 cm3 botella retornable', '$1.50')
(9, 'Pilsener 330 cm3 botella retornable', '$1.00')
(10, 'Pilsener 330 cm3 botella no retornable', '$0.90')
(11, 'Pilsener 330 cm3 Lata', '$1.30')
(12, 'Pilsener 225 cm3 botella retornable', '$0.50')
(13, 'Barriles de 30 lts', '$25.00')
(14, 'Gallito', '$0.80')
(6, 'Poni Malta - 330cm3 Lata', '$1.05')


## TAREA 1

### Insertar datos
Acceder a la BBDD facturacion e insertar un dato en la tabla producto

In [41]:
# CODE HERE

## TAREA 2

## Actualizar un dato 

Tabla producto
Vamos a actualizar la columna valor (a 1.05) en el producto de id = 6


In [ ]:
# CODE HERE

## TAREA 3

Eliminando datos!


In [48]:
# CODE HERE

(1, 1, '$3.00', 'GENERADA', 1, 1)
(2, 10, '$10.00', 'GENERADA', 1, 3)
(3, 2, '$1.20', 'GENERADA', 1, 5)
(4, 1, '$25.00', 'GENERADA', 1, 13)
(5, 1, '$0.90', 'GENERADA', 2, 10)
(6, 5, '$5.00', 'GENERADA', 2, 9)
(7, 2, '$2.60', 'GENERADA', 2, 11)
(8, 3, '$3.90', 'GENERADA', 3, 11)
(9, 1, '$0.50', 'GENERADA', 3, 12)
(10, 10, '$3.90', 'GENERADA', 4, 11)


In [126]:
# cursor  = conexion.cursor()
# cursor.execute("ROLLBACK")
# conexion.commit()

## TAREA 4

Crear una nueva tabla!
que se llame cliente 2


## Funciones

In [106]:
sql = """
    SELECT *
    FROM fn_obtener_factura_detalles(1);
"""
cursor.execute( sql )
for registro in cursor:
    print( registro )

(1, 1, '$3.00', 1, 'Coca Cola (Original) - 3lt.', '$3.00')
(2, 10, '$10.00', 3, 'Coca Cola (Original) - 1lt.', '$1.00')
(3, 2, '$1.20', 5, 'Poni Malta - 225cm3 Retornable', '$0.60')
(4, 1, '$25.00', 13, 'Barriles de 30 lts', '$25.00')


In [111]:
sql = """
    SELECT *
    FROM fn_obtener_factura_detalles(%s);
"""
valores = (1,)
cursor.execute( sql, valores )


In [112]:
df = pd.DataFrame(cursor, columns = ['id_factura_detalle_','cantidad_','cantidad_x_valor_','id_producto_','nombre_','valor_'])


In [113]:
df

id_factura_detalle_  cantidad_ cantidad_x_valor_  id_producto_  \
0                    1          1             $3.00             1   
1                    2         10            $10.00             3   
2                    3          2             $1.20             5   
3                    4          1            $25.00            13   

                          nombre_  valor_  
0     Coca Cola (Original) - 3lt.   $3.00  
1     Coca Cola (Original) - 1lt.   $1.00  
2  Poni Malta - 225cm3 Retornable   $0.60  
3              Barriles de 30 lts  $25.00

In [103]:
sql = '''select *
from factura
'''
#cursor.execute(sql)

df = pd.read_sql(sql, conexion)
df.columns

Index(['id_factura', 'fecha_emision', 'subtotal', 'valor_iva', 'valor_total',
       'estado', 'ref_empleado', 'ref_cliente'],
      dtype='object')

In [104]:
sql = """
    SELECT *
    FROM fn_obtener_factura_detalles(1);
"""
df  = pd.read_sql( sql, conexion)
df

id_factura_detalle_  cantidad_ cantidad_x_valor_  id_producto_  \
0                    1          1             $3.00             1   
1                    2         10            $10.00             3   
2                    3          2             $1.20             5   
3                    4          1            $25.00            13   

                          nombre_  valor_  
0     Coca Cola (Original) - 3lt.   $3.00  
1     Coca Cola (Original) - 1lt.   $1.00  
2  Poni Malta - 225cm3 Retornable   $0.60  
3              Barriles de 30 lts  $25.00

In [114]:
for registro in cursor:
    print( registro )

## Procedimientos

In [115]:
# CODE HERE
credenciales = {
    'dbname':'facturacion',
    'user':'postgres',
    'password':'tkmcardona',
    'host':'localhost',
    'port':5432
}

conexion = ps.connect( **credenciales )
cursor = conexion.cursor()

In [138]:
mensaje =  None
empleado =  (10,'1730581412', 'Luchito', 'Arteagada', 'lateaga', '12345', 'INSERTAR', mensaje)

cursor.execute('CALL sp_administrar_empleado(%s, %s, %s, %s, %s, %s, %s, %s)',empleado)

In [139]:

resultado = cursor.fetchone()
conexion.commit()

In [140]:
resultado[0]


'INSERTAR:duplicado'

In [141]:
sql = '''select *
from empleado
'''
#cursor.execute(sql)

df = pd.read_sql(sql, conexion)
df

id_empleado      cedula     nombre   apellido   usuario  \
0            1  0928475847      Diana    Alcazar  dalcazar   
1            2  0928387547      Lucas       Pita     lpita   
2            3  0912375847     Manuel    Salazar  msalazar   
3            4    17230398  Katherine    Morales    katymq   
4            5  1730581412    Luchito  Arteagada   lateaga   

                                               clave  
0  \xc30d040703028f10979191d085c564d23501ed81b442...  
1  \xc30d04070302b6c95d8f0c2a8f8b64d234013bd0f753...  
2  \xc30d04070302641b05dc8856d1ae61d234014debdc46...  
3  \xc30d04070302e7e4f4c25ec6403864d23501b2724e6c...  
4  \xc30d0407030239ee601bd937b5056bd236018dec00aa...

## TAREA
### ELIMINAR / ACTUALIZAR
trabajar el mismo procedimiento pero eliminado y actualizando

In [ ]:
# CODE HERE

In [21]:
# EXTRAS 
# Acceder a las tablas de la BBDD

cursor.execute("""SELECT relname FROM pg_class WHERE relkind='r'
                  AND relname !~ '^(pg_|sql_)';""") # "rel" is short for relation.

tables = [i[0] for i in cursor.fetchall()] 

tables

['proveedor',
 'proveedor_producto',
 'producto',
 'pais',
 'ciudad',
 'cliente',
 'factura',
 'empleado',
 'factura_detalle',
 'cliente2']